[2주차] 스프링을 활용한 ToDo REST API 만들기

goodcodings opened the pull request:
[오늘 한일]
1. ToDo REST API 만들기 뼈대를 만들었습니다
[내일 할일]
1. 동시성을 생각한 Task를 담을 저장소 생성
2. 요구사항 정의하기

goodcodings committed:
[2주차] 스프링을 활용한 ToDo REST API 만들기
[오늘 한일]
1. ToDo REST API 만들기 뼈대를 만들었습니다
[내일 할일]
1. 동시성을 생각한 Task를 담을 저장소 생성
2. 요구사항 정의하기

goodcodings commented:
질문이 있습니다!! 
1. lombok과 같은 외부 라이브러리를 활용해도 될까요.
2. todo를 담을 저장소는 자바 애플리케이션내 메모리로만 사용해야하나요???

johngrib commented:
> 질문이 있습니다!!
> 
> 1. lombok과 같은 외부 라이브러리를 활용해도 될까요.
> 2. todo를 담을 저장소는 자바 애플리케이션내 메모리로만 사용해야하나요???

1. 괜찮습니다. 그런데 롬복이 나오는 5강 과제부터 사용하시기를 권해드리고 싶네요.
2. 테스트 가능한 형태라면  그 외의 방법도 괜찮습니다만, 과정에서 너무 벗어나지 않도록 주의해 주세요. 일단은 지금 과정에서의 제약 사항이라고 생각해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/common/response/CommonResponse.java:
```diff
@@ -0,0 +1,112 @@
+import com.codesoom.assignment.common.Builder;
+import com.codesoom.assignment.common.response.messages.error.ErrorCode;
+
+public class CommonResponse {
```
이 클래스에 JavaDoc 주석을 달아서 의도와 용도에 대해 설명해 보세요.

goodcodings replied:
해당 클래스는 요청에 대한 응답의 일괄성을 위한 역활을 위임 받은 클래스 입니다!
JavaDoc에 더 자세한 설명을 남기겠습니다!

goodcodings commented:
> > 질문이 있습니다!!
> > 
> > 1. lombok과 같은 외부 라이브러리를 활용해도 될까요.
> > 2. todo를 담을 저장소는 자바 애플리케이션내 메모리로만 사용해야하나요???
> 
> 1. 괜찮습니다. 그런데 롬복이 나오는 5강 과제부터 사용하시기를 권해드리고 싶네요.
> 2. 테스트 가능한 형태라면  그 외의 방법도 괜찮습니다만, 과정에서 너무 벗어나지 않도록 주의해 주세요. 일단은 지금 과정에서의 제약 사항이라고 생각해 주세요.

넵!!  과정에 맞춰 과제를 수행하겠습니다!!!

johngrib commented on app/src/main/java/com/codesoom/assignment/common/response/CommonResponse.java:
```diff
@@ -0,0 +1,112 @@
+
+public class CommonResponse {
+    private Result result;
+    private Object data;
```
제네릭을 쓰는 방법도 고려해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/common/response/messages/error/ErrorCode.java:
```diff
@@ -0,0 +1,20 @@
+package com.codesoom.assignment.common.response.messages.error;
+
+public enum ErrorCode {
```
내용을 보면 좀 더 구체적인 이름을 사용해도 괜찮을 것 같군요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,87 @@
+        this.taskId = taskBuilder.taskId;
+    }
+
+    public Task registerTask() {
+        this.status=Status.INCOMPLETE;
+        this.createdAt = LocalDateTime.now();
+        this.updatedAt = LocalDateTime.now();
+        this.taskId = UUID.randomUUID().toString()
+                .replaceAll("-", "")
+                .substring(0, 8);
+        return this;
+    }
```
흥미롭게 잘 만드신 것 같습니다. 이런 종류의 구현을 했을 때에는 리뷰어에게 먼저 설명을 하는 습관을 들이는 것이 좋습니다. (제가 지금 이 코드에 리뷰를 남기는 것처럼) 이병성 님도 이 코드에 스스로 리뷰를 남겨서 이 코드가 어떤 의미가 있는지 설명해 보세요. 6줄의 코드를 작성하면서도 많은 고민과 결정을 내리셨을 것으로 생각합니다. 그 중에서 핵심적인 아이디어를 설명하면 돼요.

예를 들어서 `.substring(0,8)`하나에 대해서도 taskId 의 공간 규모가 어느 정도가 되는지를 설명할 수 있겠죠. 개발자에게 중요한 스킬인 큰 수의 추정을 연습해볼 수 있는 기회입니다. 

막연한 것 같다면 이런 식으로 시작할 수 있을 겁니다. "taskId의 공간이 16^8 이므로, 몇 명의 사용자가 몇 개의 할 일을 남길 수 있고, 이는 어떠어떠한 것을 전제한다면 서비스를 몇 년간 운영할 경우 공간이 바닥나게 된다(또는 그 이후에도 넉넉할 것으로 예상한다).. 그리고..."

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,87 @@
+        return this;
+    }
+
+    /*
+         task아이디를 통헤서 기존에 등록된 task를 가져온 후
+         title과 update시간을 변경할 예정입니다.
+     */
```
이렇게 할 일을 주석으로 남기는 경우에는 `TODO` 태그를 입력해 주세요. 찾기도 쉽고 IDE도 관리를 도와줄 것입니다.

```suggestion
    /*
         TODO: task아이디를 통헤서 기존에 등록된 task를 가져온 후
         title과 update시간을 변경할 예정입니다.
     */
```

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/TaskCommand.java:
```diff
@@ -0,0 +1,80 @@
+import com.codesoom.assignment.common.response.messages.error.ErrorCode;
+import org.springframework.util.Assert;
+
+/*
+    외부에서 직접적으로 엔티티 개념인 Task를 접근하지 않기 위해 작성한 인터페이스 객체입니다
+ */
```
JavaDoc의 표준적인 포맷이 어떻게 되는지를 조사해 보시고, 적용해 보세요. 공부한 내용을 글로 적어서 공유해보면 더 좋겠죠!

johngrib commented on app/src/main/java/com/codesoom/assignment/interfaces/TodoDto.java:
```diff
@@ -0,0 +1,52 @@
+
+/*
+    Todo를 등록,삭제,수정등 클라이언트로 부터 전달 받은 데이터를 각 액션마다  Paramter Object를 만들었습니다.
+    인터넷 매체에 많은 정보가 존재하지 않아 리팩토링에서 제시한  Paramter Object를 제대로 구현했는지 확신이 서지 않습니다..ㅠㅠ
```
왜 Parameter Object 패턴을 사용하는 것이 필요하다고 생각하셨는지를 먼저 설명해 보세요.

goodcodings replied:
등록, 삭제, 수정, 상태 변경의 요청이 들어올 경우 도메인쪽으로 데이터를 넘겨 줄때 
매개변수로만 넘겨주게 되면 추후에 요구사항이 변경될 경우 컨트롤러와 도메인 객체로 변환해주는 두 클래스 모두 변경될 가능성이 있어 OCP에 위반이 될거라 생각해서

파라미터 오브젝트 패턴을 사용해야겠다고 생각했습니다!

johngrib replied:
좋습니다! 좀 쑥쓰럽고 어려울 수도 있지만, 항상 언제든지 자신의 코드를 설명할 수 있도록 연습해 보세요! 혼자서 코딩할 때에도 남에게 설명한다고 생각하고 글을 작성해 보거나 혼잣말로 왜 이렇게 했는지 이야기해보는 시간을 가져 보세요.

johngrib reviewed (commented):
반갑습니다 :-)

goodcodings committed:
[2주차] 스프링을 활용한 ToDo REST API 만들기
[오늘 한일]
1. JavaDoc를 부분으로 적용했습니다
[내일 할일]
1. JavaDoc마저 적용하기
2. 동시성을 생각한 Task를 담을 저장소 생성
3. 요구사항 정의하기

goodcodings commented:
적은 부분이지만 개발 보단 부분적으로 JavaDoc적용을 시작했습니다..
처음 적용해보니 클래스 및 메서드에 대한 설명을 어떻게 작성해야하는지 생각이 바로 나지 않네요!
연습을 더 많이 해서 간결하고 명확한 설명을 작성할 수 있도록 노력하겠습니다!

감사합니다!

goodcodings commented on app/src/main/java/com/codesoom/assignment/common/response/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,19 @@
+
+
+@ControllerAdvice
+public class ControllerExceptionHandler {
```
해당 클래스는 예외 발생할 경우 통합으로 관리하기 위한 클래스입니다!

johngrib replied:
이런 설명을 JavaDoc으로 작성하시면 됩니다 ㅎㅎ

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -5,23 +5,48 @@
 import java.time.LocalDateTime;
 import java.util.UUID;
 
+/**
+ * task란 도메인을 개념적으로 표현한 클래스입니다.
+ *
+ * @author Lee Byoung Seong
+ * @version 1.0
+ */
```
JavaDoc을 처음 작성하시는 것인데도 용감하게 잘 작성하셨습니다. 몇 가지 신경써야 하는 것이 있으니 고려해서 업데이트해보세요!

- `task란 도메인을 개념적으로 표현한 클래스입니다.`
    - `task`를 여기에 써버리면 클래스 이름을 바꿀 일이 생겼을 때 주석도 고쳐줘야 하는 문제가 생깁니다. 객체와 객체 사이의 커플링이 주석과 코드 사이에서도 발생하는 경우라 할 수 있습니다.
- `@author` 요즘은 이걸 굳이 적어줄 필요가 없습니다. git에 기록이 다 남기도 하고요. (Java 빌트인 패키지 같은 경우는 예외입니다)
- `@version` 이번 프로젝트에서는 생략해 주셔도 됩니다. JavaDoc에서 제일 중요한 건 메인 설명문이니 코드숨에서는 거기에 좀 더 집중하도록 합시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -5,23 +5,48 @@
-    private enum Status{
-        REMOVE,INCOMPLETE,COMPLETE
+    /**
+     * Task의 상태를 그룹화한 enum입니다.
+     * <p>
+     * {@code REMOVE} 삭제된 상태
+     * {@code INCOMPLETE} 미완료 상태
+     * {@code COMPLETION} 완료 상태
+     */
+    private enum Status {
+        REMOVE, INCOMPLETE, COMPLETION
```
위의 리뷰를 읽고 곰곰히 생각해보시면 이 JavaDoc에 어떤 문제가 있는지 짐작하실 수 있을 것입니다. 어떤 문제가 있는지 설명해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/interfaces/TodoDto.java:
```diff
@@ -1,52 +1,137 @@
-    Todo를 등록,삭제,수정등 클라이언트로 부터 전달 받은 데이터를 각 액션마다  Paramter Object를 만들었습니다.
-    인터넷 매체에 많은 정보가 존재하지 않아 리팩토링에서 제시한  Paramter Object를 제대로 구현했는지 확신이 서지 않습니다..ㅠㅠ
+/**
+ * 클라이언트의 각 요청(등록,수정,삭제,완료)에 대응하는 정적멤버클래스가 존재하는 클래스입니다.
```
클래스에 대해 설명할 때 `얘는 클래스입니다` 라고 설명하지 않아도 됩니다. 잘 모르겠다면 이 클래스의 **책임**이 무엇인지 생각해서 무엇을 책임지는지에 대해 작성해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/interfaces/TodoDto.java:
```diff
@@ -1,52 +1,137 @@
         private final String title;
 
+        /**
+         * 제목을 초기화 시킵니다
```
`초기화시킨다`는 표현은 애매합니다. 구체적이고 분명한 표현을 사용해 봅시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/interfaces/TodoDto.java:
```diff
@@ -1,52 +1,137 @@
+         * 제목을 초기화 시킵니다
+         *
+         * @param title 제목
+         * @throws IllegalArgumentException {@code 제목}이 존재하지 않으면 던집니다.
```
여기에 `@code`를 사용하지 않아도 괜찮습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/interfaces/TodoDto.java:
```diff
@@ -1,52 +1,137 @@
         }
 
+        /**
+         * 제목을 건내 줍니다.
```
`건네 준다`는 표현은 애매합니다.

`String.java`을 열어보시고, JavaDoc이 어떻게 작성되어 있는지 한번 살펴봅시다. 

![image](https://user-images.githubusercontent.com/1855714/160834115-502563c4-3fb9-4e81-b2f7-1e611804557d.png)

"이 문자열의 길이를 리턴합니다." 라고 되어 있죠. JavaDoc에서 가장 중요한 것은 이 메소드가 "무엇을 리턴하는지 설명하는 것"이라 해도 과언이 아닙니다. 이제 이 JavaDoc 설명문을 수정해 봅시다.

johngrib reviewed (commented):
JavaDoc 화이팅!

goodcodings commented:
꼼꼼한 피드백을 해주셔서 감사합니다!!!
종립님이 작성하신 JavaDoc에 대한 사용법과 String과 ArrayList를 보면서 제꺼로 만들어 봤는데 
아직 부족한점이 많네요!!

Lee-Byoung-Seong committed:
[2주차] 스프링을 활용한 ToDo REST API
[오늘 한 일]
종립님으로 부터 피드백 받은 JavaDoc내용을 수정했습니다

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,106 @@
+import java.time.LocalDateTime;
+import java.util.UUID;
+
+/**
+ *  할일 도메인을 개념적으로 표현한 엔티티입니다.
+ */
```
이렇게 하면 어떨까요?
```suggestion
/**
 *  할 일.
 */
```

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,106 @@
+    private LocalDateTime updatedAt;
+    private Status status;
+
+    /**
+     * 할일 도메인의 상태를 그룹화한 enum입니다.
+     */
```
```suggestion
    /**
     * 할 일의 상태.
     */
```

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,106 @@
+        REMOVE, INCOMPLETE, COMPLETION
+    }
+
+    /**
+     * 전달 받은 아이디와 제목을 초기화 시킵니다.
+     *
+     * @param taskBuilder
+     */
```
이렇게 하면 다른 필드들이 추가될 때마다 주석도 계속 바뀌겠죠. 이렇게 해보세요.
```suggestion
    /**
     * 할 일을 생성합니다.
     *
     * @param taskBuilder
     */
```

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,106 @@
+
+    /**
+     * 할일 등록시 상태, 등록일자, 수정일자, 아이디를 초기화 시킵니다.
+     * 객체의 상태를 스스로 책임시키기 위해 존재합니다
```
이 설명은 미묘합니다. 기왕 설명하려면 더 자세히 설명해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/TaskCommand.java:
```diff
@@ -0,0 +1,80 @@
+import org.springframework.util.Assert;
+
+/**
+ *   클라이언트로 전달받은
```
여기는 문장이 완성되지 않았습니다.

johngrib commented:
한 주간 고생 많으셨습니다.
